base/strings/substring.jl

```diff
@@ -55,7 +55,8 @@ String(s::SubString{String}) = unsafe_string(pointer(s.string, s.offset+1), s.nc
 
 ncodeunits(s::SubString) = s.ncodeunits
 codeunit(s::SubString) = codeunit(s.string)
-length(s::SubString) = length(s.string, s.offset+1, s.offset+s.ncodeunits)
+length(s::SubString) =
+    s.ncodeunits == 0 ? 0 : length(s.string, s.offset+1, s.offset+s.ncodeunits)
```

Comment:
Do you really need the conditional? I don't think you do since e.g. `length("Hello", 3, 2) == 0`.

Reply:
The problem is that:
```
julia> length("Hello", 10, 9)
ERROR: BoundsError: attempt to access "Hello"
  at index [10]
Stacktrace:
 [1] length(::String, ::Int64, ::Int64) at .\strings\string.jl:242
 [2] top-level scope
```

I have either to:
1. check here
2. change `SubString` constructor to always make `offset` of empty string equal to `0`
3. change `length` method for string not to throw error on invalid empty values of `i` and `j`.

Which one do you prefer?

Reply:
I guess I kind of think that `"Hello"[10:9]` and `SubString("Hello", 10, 9)` should be errors in the first place now that I think about it. I suspect that for `SubString(s, i, j)` the bounds on `i` should be `0 < i ≤ ncodeunits(s)+1` and the bounds on `j` should be `0 ≤ j < ncodeunits(s)+1`. This is, incidentally I think exactly the bounds on `i` and `j` in `length(s, i, j)` so they line up perfectly. That expands the scope of this a bit but I think it's the right solution.

Reply:
I like this approach, but before I implement let me verify with you one assumption. Which approach to checking of validity of indices we want to take:
1. we allow `i` and `j` to be invalid, e.g. `"∀"[2:1]` and only check if they are in range?
2. we allow only `j` to be invalid, but `i` must be a valid index?

And yes - this will be a significant, breaking change, a lot of test code has to be rewritten apart from constructors and `getindex`.

Additionally (this may be a separate PR, but relates to bounds checking so allow me clarify it here). Is this a desired behavior (i.e. to get negative output from length):
```
julia> [length("∀", i, j) for i in 1:4, j in 0:3]
4×4 Array{Int64,2}:
 0  1   1   1
 0  0  -1  -1
 0  0   0  -2
 0  0   0   0
```

Reply:
Yikes, those negative values are not intended. I've been meaning to go back and write a bunch of tests for this code, but the code loading changes have dragged on forever so I've been stuck working on that instead. I think the indices for `SubString` should be required to be in-bounds and valid or be `i == ncodeunits(s)+1` or `j == 0`. It may also be a good idea to throw an error for `j - i + 1 < 0` – i.e. allow the empty case for string slicing but not the "negative length" cases.

For the `length` function, it needs to at least support every set of indices that `SubString` does, but I think it can be a little bit more lenient, e.g. allowing `i` and `j` to be invalid indices. Whether we should allow `j - i + 1 < 0` for `length` and just return `0` or not, I'm not sure. It's probably not a bad idea.

Reply:
I propose to do the following:
1. first I fix `length` in a separate PR;
2. when `length` PR is merged I rebase this PR and do the following (meantime maybe we get some feedback on those assumptions as they are breaking):
    * for all `<:AbstractString` in base check if `i` is a valid index or is `ncodeunits(s)+1` in `getindex` no matter what `j` is passed; THIS IS BREAKING
    * for `SubString` constructor check the same as above; THIS IS BREAKING
    * I would allow `j` to take any value, as `i-1` does not have to be valid (some smaller `j` may be `prevind`) and using `j=0` is also a natural choice to render an empty string.